src/cargo/core/compiler/fingerprint.rs

```diff
@@ -212,28 +213,49 @@ enum LocalFingerprint {
 }
 
 impl LocalFingerprint {
-    fn mtime(root: &Path, mtime: Option<FileTime>, path: &Path) -> LocalFingerprint {
+    fn mtime(root: &Path, mtime: MtimeState, path: &Path) -> LocalFingerprint {
         let mtime = MtimeSlot(Mutex::new(mtime));
         assert!(path.is_absolute());
         let path = path.strip_prefix(root).unwrap_or(path);
         LocalFingerprint::MtimeBased(mtime, path.to_path_buf())
     }
 }
 
-struct MtimeSlot(Mutex<Option<FileTime>>);
+struct MtimeSlot(Mutex<MtimeState>);
+
+#[derive(Debug)]
+enum MtimeState {
+    LatestTimeIs(FileTime),
+    OlderThan(FileTime),
+    NotAvailable,
+}
 
 impl Fingerprint {
     fn update_local(&self, root: &Path) -> CargoResult<()> {
         let mut hash_busted = false;
         for local in self.local.iter() {
-            match *local {
-                LocalFingerprint::MtimeBased(ref slot, ref path) => {
-                    let path = root.join(path);
+            let (slot, path) = match local {
+                LocalFingerprint::MtimeBased(slot, path) => (slot, path),
+                LocalFingerprint::EnvBased(..) |
+                LocalFingerprint::Precalculated(..) => continue,
+            };
+            let path = root.join(path);
+            let mut slot = slot.0.lock().unwrap();
+            *slot = match &*slot {
+                MtimeState::OlderThan(newest_file) => {
+                    let nanos = newest_file.nanoseconds() + 1;
+                    let time_to_set = FileTime::from_unix_time(
+                        newest_file.seconds() + (nanos as i64) / 1_000_000_000,
+                        nanos % 1_000_000_000,
+                    );
+                    filetime::set_file_times(&path, time_to_set, time_to_set)?;
+                    MtimeState::LatestTimeIs(time_to_set)
+                }
+                _ => {
```

Comment:
This is kind of general, but I tend to avoid using default cases on sum types because if that algebra every changes (in this case, say an `OlderThanOrEqual` case is added) it's easy to miss changing these code paths because the exhaustivity checker won't tell you.

In short I'd use

```rust
MtimeState::LatestTimeIs(_) | NotAvailable => {
```

here instead.

Good practice or YAGNI? 🤷‍♂️

Reply:
Sounds good to me!